File: Cpp/1030_Matrix_Cells_in_Distance_Order.cpp
class Solution {
public:
    vector<vector<int>> allCellsDistOrder(int R, int C, int r0, int c0) {
        vector<vector<int>> result;
        vector<vector<bool>> visit(R, vector<bool>(C, false));
        
        queue<pair<int, int>> q;
        q.push({r0, c0});
        
        while(!q.empty()){
            auto cell = q.front();
            q.pop();
            
            visit[cell.first][cell.second] = true;
            result.push_back({cell.first, cell.second});
            
            if(cell.first-1>=0 && !visit[cell.first-1][cell.second]) q.push({cell.first-1, cell.second});
            if(cell.first+1<R && !visit[cell.first+1][cell.second]) q.push({cell.first+1, cell.second});
            if(cell.second-1>=0 && !visit[cell.first][cell.second-1]) q.push({cell.first, cell.second-1});
            if(cell.second+1<C && !visit[cell.first][cell.second+1]) q.push({cell.first, cell.second+1});
        }
        
        return result;
    }
};

File: Kotlin/1030_Matrix_Cells_in_Distance_Order.kt
class Solution {
    fun allCellsDistOrder(rows: Int, cols: Int, rCenter: Int, cCenter: Int): Array<IntArray> {
        val result = Array (rows * cols) { IntArray(2) }
        val visit = Array(rows) { BooleanArray(cols) }
        val q = LinkedList<Pair<Int,Int>>()

        q.push(Pair(rCenter, cCenter))
        
        var index = 0
        while(!q.isEmpty()) {
            val cell = q.remove()
            if (visit[cell.first][cell.second]) continue
            
            result[index][0] = cell.first
            result[index++][1] = cell.second
            
            visit[cell.first][cell.second] = true
            
            if(cell.first - 1 >= 0 && !visit[cell.first - 1][cell.second]) q.add(Pair(cell.first - 1, cell.second))
            if(cell.first + 1 < rows && !visit[cell.first + 1][cell.second]) q.add(Pair(cell.first + 1, cell.second))
            if(cell.second - 1 >= 0 && !visit[cell.first][cell.second - 1]) q.add(Pair(cell.first, cell.second - 1))
            if(cell.second + 1 < cols && !visit[cell.first][cell.second + 1]) q.add(Pair(cell.first, cell.second + 1))
        }
        
        return result
    }
}